dts: Add support for io-channel properties

jimparis opened the pull request:
This adds support for e.g. `io-channels = <&adc 3>` analogous to existing `pwms = <&pwm 0 10000>`, for referencing a specific IO channel input.  Also adds support to the `nrf-saadc` bindings.  This uses the new shared logic from #18065 

Example of how I'm using it:

`dts/bindings/adc-map.yaml`:
```
title: ADC Map

description: >
    Map ADC input numbers to names

inherits:
    !include base.yaml

properties:
    compatible:
      constraint: "adc-map"

sub-node:
  properties:
    io-channels:
      type: compound
      category: required
```

`boards/arm/foo/foo.dts`:
```
/dts-v1/;
#include <nordic/nrf52832_qfaa.dtsi>

/ {
        model = "Foo";
        compatible = "nordic,nrf52832-qfaa", "nordic,nrf52832";

        adc-map {
                compatible = "adc-map";
                batt_meas {
                        io-channels = <&adc 0>;
                };
        };
};

&adc {
        status = "okay";
        #io-channel-cells = <1>;
};
```

`app/power.c`:
```c
adc_dev = device_get_binding(DT_ADC_MAP_BATT_MEAS_IO_CHANNELS_CONTROLLER);
const struct adc_channel_cfg cfg = {
        .gain             = ADC_GAIN_1_6,
        .reference        = ADC_REF_INTERNAL,
        .acquisition_time = ADC_ACQ_TIME(ADC_ACQ_TIME_MICROSECONDS, 10),
        .channel_id       = 0,
        .input_positive   = (NRF_SAADC_INPUT_AIN0 + DT_ADC_MAP_BATT_MEAS_IO_CHANNELS_INPUT),
};
adc_channel_setup(adc_dev, &cfg);
```

jimparis committed:
dts/bindings: add '#cells' to nrf-saadc

This is to support e.g. "<&adc 3>" in the device tree to create e.g.

  DT_FOO_IO_CHANNELS_CONTROLLER = "ADC_0"
  DT_FOO_IO_CHANNELS_INPUT = 3

Signed-off-by: Jim Paris <jim@jtan.com>

ulfalizer reviewed (approved):
Fine implementation-wise to me. Other people will have to comment on other stuff.

Some documentation outside the code might be nice at some point too.

galak reviewed (commented):
Take a look at:

https://git.kernel.org/pub/scm/linux/kernel/git/devicetree/devicetree-rebasing.git/tree/Bindings/iio/iio-bindings.txt

Can we use these semantics to accomplish what you want?  Looks similar with different names.

jimparis commented:
> Take a look at:
> 
> https://git.kernel.org/pub/scm/linux/kernel/git/devicetree/devicetree-rebasing.git/tree/Bindings/iio/iio-bindings.txt
> 
> Can we use these semantics to accomplish what you want? Looks similar with different names.

Yeah, I think we'd be pretty close if we named it `io-channels` and I'm not wedded to `adcs` if we want to change it.  `io-channels` seems to cover more than ADCs, though, and I don't know how we'd support different things (my current C code of course assumes `*_CONTROLLER` is an ADC device).

galak commented:
> Yeah, I think we'd be pretty close if we named it `io-channels` and I'm not wedded to `adcs` if we want to change it. `io-channels` seems to cover more than ADCs, though, and I don't know how we'd support different things (my current C code of course assumes `*_CONTROLLER` is an ADC device).

I think its fine, I'd prefer we try and match existing standard for DT.  So let's use `io-channels`

jimparis committed:
scripts/dts: Support 'io-channels' property just like 'pwms'

This is a direct search-and-replace copy of the PWM code.
The name is chosen to match Linux's iio-bindings.txt.

Signed-off-by: Jim Paris <jim@jtan.com>

jimparis committed:
scripts/dts: add basic test for IO channel property

Test the new io-channel support

Signed-off-by: Jim Paris <jim@jtan.com>

jimparis commented:
Updated from `adcs` to `io-channels`

ulfalizer commented on scripts/dts/edtlib.py:
```diff
@@ -243,8 +244,7 @@ def __repr__(self):
 class Device:
-    """
-    Represents a device. There's a one-to-one correspondence between device
+    """Represents a device. There's a one-to-one correspondence between device
```
Accidental change? The newline was deliberate.

jimparis replied:
yeah I'll have to blame emacs `fill-paragraph` for that one

ulfalizer commented on scripts/dts/edtlib.py:
```diff
@@ -340,6 +345,7 @@ class Device:
     flash_controller:
       The flash controller for the device. Only meaningful for devices
       representing flash partitions.
+
```
Accidental change?

ulfalizer commented on scripts/dts/edtlib.py:
```diff
@@ -950,6 +962,39 @@ def __repr__(self):
 
+class IOChannel:
+    """
+    Represents an IO channel used by a device.
```
Maybe a link to that Linux kernel document would be nice here.

ulfalizer reviewed (changes requested):
Just some nits.

galak commented:
Do we need to update all the adc's to have '#io-channel-cells' and such?

jimparis commented:
Probably?  I wasn't sure what hardware supported this concept and which didn't.  If it's all of them, it would be nice to put this in the base `adc.yaml` instead, but that didn't seem to work when I tried it.

galak commented:
> Probably? I wasn't sure what hardware supported this concept and which didn't. If it's all of them, it would be nice to put this in the base `adc.yaml` instead, but that didn't seem to work when I tried it.

Lets work on that as a follow up to this.

pabigot commented:
@jimparis @galak This merge included a change to `nordic,nrf-saadc` to add a `#cells` property.  Was that necessary?  If so, don't we also need it in `nordic,nrf-adc` and probably all the other ADC peripheral descriptions?

If not, can we get rid of it?

galak commented:
> @jimparis @galak This merge included a change to `nordic,nrf-saadc` to add a `#cells` property. Was that necessary? If so, don't we also need it in `nordic,nrf-adc` and probably all the other ADC peripheral descriptions?
> 
> If not, can we get rid of it?

See my new PR: #18589
